### PR TITLE
[SYCL] Implement group_barrier for group class

### PR DESCRIPTION
**Pull request by alexbatashev**

### CODE REVIEW

File: sycl/include/CL/sycl/group.hpp [hunk 1]
```diff
@@ -435,5 +451,18 @@ template <int Dims> group<Dims> this_group() {
...
+}
+template <>
+inline void group_barrier<group<2>>(group<2>, memory_scope FenceScope) {
+  detail::workGroupBarrier(FenceScope);
+}
+template <>
+inline void group_barrier<group<3>>(group<3>, memory_scope FenceScope) {
+  detail::workGroupBarrier(FenceScope);
+}
+
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/group.hpp:466]: Could you please clarify why 3 identical specializations are needed? Can we just provide definition of declaration on line 452?

File: sycl/include/CL/sycl/group.hpp [hunk 2]
```diff
@@ -40,6 +41,21 @@ static inline void workGroupBarrier() {
...
 
+static inline void workGroupBarrier(memory_scope FenceScope) {
+  (void)FenceScope;
+#ifdef __SYCL_DEVICE_ONLY__
+  __spirv_ControlBarrier(__spv::Scope::Workgroup,
+                         sycl::detail::spirv::getScope(FenceScope),
+                         __spv::MemorySemanticsMask::AcquireRelease |
+                             __spv::MemorySemanticsMask::SubgroupMemory |
+                             __spv::MemorySemanticsMask::WorkgroupMemory |
+                             __spv::MemorySemanticsMask::CrossWorkgroupMemory);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/group.hpp:52]: Could you please add a comment clarifying why  this set of flags in required?

File: sycl/include/CL/sycl/group_algorithm.hpp [hunk 3]
```diff
@@ -834,5 +834,36 @@ joint_inclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
       sycl::known_identity_v<BinaryOperation, typename OutPtr::element_type>);
 }
 
+namespace detail {
+template <typename G> struct group_barrier_scope {};
+template <> struct group_barrier_scope<sycl::sub_group> {
+  constexpr static auto scope = __spv::Scope::Subgroup;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/group_algorithm.hpp:840]:
```suggestion
  constexpr static auto Scope = __spv::Scope::Subgroup;
```

File: sycl/include/CL/sycl/group_algorithm.hpp [hunk 4]
```diff
@@ -834,5 +834,36 @@ joint_inclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
...
+};
+template <int D> struct group_barrier_scope<sycl::group<D>> {
+  constexpr static auto scope = __spv::Scope::Workgroup;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/group_algorithm.hpp:843]:
```suggestion
  constexpr static auto Scope = __spv::Scope::Workgroup;
```

File: sycl/include/CL/sycl/group_algorithm.hpp [hunk 5]
```diff
@@ -834,5 +834,36 @@ joint_inclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
...
+};
+} // namespace detail
+
+template <typename Group>
+typename std::enable_if<is_group_v<Group>>::type
+group_barrier(Group, memory_scope FenceScope = Group::fence_scope) {
+  (void)FenceScope;
+#ifdef __SYCL_DEVICE_ONLY__
+  // Per SYCL spec, group_barrier must perform both control barrier and memory
+  // fence operations. All work-items execute release a release fence prior to
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/group_algorithm.hpp:853]:
```suggestion
  // fence operations. All work-items execute a release fence prior to
```
?

File: sycl/test/check_device_code/group_barrier.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    CGH.parallel_for(sycl::nd_range{sycl::range{1, 1}, sycl::range{1, 1}},
+                     [](sycl::nd_item<2> item) {
+                       auto G = item.get_group();
+                       TestLambda(G);
+                     });
+  });
+  Q.submit([](sycl::handler &CGH) {
+    CGH.parallel_for(sycl::nd_range{sycl::range{1, 1, 1}, sycl::range{1, 1, 1}},
+                     [](sycl::nd_item<3> item) {
+                       auto G = item.get_group();
```
Comment (romanovvlad) [inline sycl/test/check_device_code/group_barrier.cpp:33]: Could you please add a test with  sub_group's?

Comment (alexbatashev) [reply sycl/test/check_device_code/group_barrier.cpp:33]: There's already a test in llvm-test-suite. But I can add one more.

File: sycl/include/CL/sycl/group_algorithm.hpp [hunk 7]
```diff
@@ -834,5 +834,36 @@ joint_inclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
...
+  // barrier and acquire fence afterwards. The rest of semantics flags specify
+  // which type of memory this behavior is applied to.
+  __spirv_ControlBarrier(detail::group_barrier_scope<Group>::scope,
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/group_algorithm.hpp:856]:
```suggestion
  __spirv_ControlBarrier(detail::group_barrier_scope<Group>::Scope,
```

File: sycl/test/check_device_code/group_barrier.cpp (hunk 6 above)
Comment (romanovvlad) [reply sycl/test/check_device_code/group_barrier.cpp:33]: It would be great.

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**